service/storage_service.cc

```diff
@@ -5198,13 +5200,27 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     //co_await sleep_abortable(_db.local().get_config().read_request_timeout_in_ms() * std::chrono::milliseconds(1), _abort_source);
                     result.status = raft_topology_cmd_result::command_status::success;
                 break;
+                case raft_topology_cmd::command::fence: {
+                    const auto version = _topology_state_machine._topology.version;
+                    co_await update_fence_version(version);
```

Comment:
No, it should be run sequentially by the topology change coordinator.

Reply:
What if there's coordinator failover?

Reply:
It seems nothing prevents a race:
* coordinator A with fence_version=1 reached `update_fence_version`
* coordinator B with fence_version=2 came
* raft already updated its term, so it's matched
* coordinator B finishes `update_fence_version`, checks pass since fence_version=2 > 1 - the previous version
* coordinator B finishes `_sys_ks.local().update_topology_fence_version`;
* coordinator A finishes `_sys_ks.local().update_topology_fence_version`, overwriting fence_version back to 1.

We should probably grab some `semaphore` before checking term and index.

Reply:
The spec says:

```
def fence(v):
   set the minimal accepted token metadata version of the fence to v, \
        unless the fence already rejects v
```

So shouldn't we just ignore the update that tries to move the fence back?

Reply:
> So shouldn't we just ignore the update that tries to move the fence back?

How to guarantee atomicity between these two lines:
```
co_await update_fence_version(version);
co_await _sys_ks.local().update_topology_fence_version(version);
```

Reply:
> > So shouldn't we just ignore the update that tries to move the fence back?
> 
> How to guarantee atomicity between these two lines:
> 
> ```
> co_await update_fence_version(version);
> co_await _sys_ks.local().update_topology_fence_version(version);
> ```

To not override newer version with the older one in the persistent storage we can use timestamps. The newer write will use larger timestamp. It would have been easy if we would use timed uuid as the version identifier, but I think we can have a way to generate monotonic timestamp without it as well.

Reply:
> To not override newer version with the older one in the persistent storage we can use timestamps.

I’m not sure I understand. Are you proposing some checks in CQL, like `update scylla_local set value = x where key = y and value value < x`? Now it’s `set_scylla_local_param_as`, we will need to refactor it. 

A lock looks simpler:
```
case raft_topology_cmd::command::fence: {
    // just a single line, update_fence_version throws if version < current_version
    auto operation_holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);

    const auto version = _topology_state_machine._topology.version;
    co_await update_fence_version(version);
    co_await _sys_ks.local().update_topology_fence_version(version);
    result.status = raft_topology_cmd_result::command_status::success;
    break;
}
```

Reply:
> > To not override newer version with the older one in the persistent storage we can use timestamps.
> 
> I’m not sure I understand. Are you proposing some checks in CQL, like `update scylla_local set value = x where key = y and value value < x`? Now it’s `set_scylla_local_param_as`, we will need to refactor it.
> 

No. I propose to do `UPDATE scylla_local ... USING TIMESTAMP X` where a write with smaller version will have smaller timestamp. The visible write is the one with the newest timestamp.
 
> A lock looks simpler:
> 
> ```
> case raft_topology_cmd::command::fence: {
>     // just a single line, update_fence_version throws if version < current_version
>     auto operation_holder = co_await get_units(_raft_topology_cmd_handler_state._operation_mutex, 1);
> 
>     const auto version = _topology_state_machine._topology.version;
>     co_await update_fence_version(version);
>     co_await _sys_ks.local().update_topology_fence_version(version);
>     result.status = raft_topology_cmd_result::command_status::success;
>     break;
> }
> ```

A lock may work too, but timestamp is the usual way it is done lockless in Cassandra/Scylla.